### PR TITLE
Sync user info

### PR DESCRIPTION
**Pull request by GDay**
This new feature will allow users to sync data from a third party back to the ChiefOnboarding users. It will require a new integration and will run daily. 

This is especially useful if you need to call API endpoints in any of your integrations and require the ID of the user, but the API doesn't give you an option to look that ID up (for example through the email address). 

The data is saved in the `extra_fields` field and can be used in any of the integrations and content boxes.

### CODE REVIEW

File: back/admin/integrations/serializers.py [hunk 1]
```diff
@@ -119,12 +119,6 @@ class ManifestOauthSerializer(ValidateMixin, serializers.Serializer):
 
 
 class ManifestSerializer(ValidateMixin, serializers.Serializer):
-    type = serializers.ChoiceField(
-        [
-            ("import_users", "imports users from endpoint"),
```
Comment (GDay) [inline back/admin/integrations/serializers.py:124]: Unrelated change. I forgot to remove this part after making this a field in the integration object.

File: back/admin/integrations/sync_userinfo.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+
+class SyncUsers(PaginatedResponseMixin):
+    """
+    Sync some data from a third party to our users. This can be anything that is
+    presented in a list form from the users. Matching happens based on the email
+    property from the response data. Paginated response is supported.
+    """
+
+    def __init__(self, integration):
```
Comment (cscheng) [inline back/admin/integrations/sync_userinfo.py:18]: Doesn't look like it's overriding the super `__init__()` or doing something different.

File: back/admin/integrations/import_users.py [hunk 3]
```diff
@@ -1,128 +1,20 @@
...
+class ImportUser(PaginatedResponseMixin):
     """
-    Extension of the `Integration` model. This part is only used to get users
-    from a third party API endpoint and format them in a way that we can proccess
-    them.
+    Extracts all users and then checks which ones are not yet in the system,
+    so we can show them as options when importing these.
     """
 
     def __init__(self, integration):
```
Comment (cscheng) [inline back/admin/integrations/import_users.py:13]: Doesn't look like it's different from parent class init method.

File: back/admin/integrations/import_users.py [hunk 4]
```diff
@@ -1,128 +1,20 @@
...
 
-from admin.integrations.utils import get_value_from_notation
-from django.utils.translation import gettext_lazy as _
-from admin.integrations.exceptions import GettingUsersError
+from admin.integrations.mixins import PaginatedResponseMixin
+from organization.models import Organization
 
 
-class ImportUser:
+class ImportUser(PaginatedResponseMixin):
```
Comment (cscheng) [inline back/admin/integrations/import_users.py:7]: Should this be ImportUsers (plural) as the other integration is called SyncUsers?

File: back/admin/integrations/sync_userinfo.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        users = self.get_data_from_paginated_response()
+
+        # get user emails so we can fetch all users from the db (without having to
+        # make a bunch of queries). Email param is currently hardcoded, no way to change
+        emails = [user["email"] for user in users]
+
+        user_objects = get_user_model().objects.filter(email__in=emails)
+
+        for user in user_objects:
+            user_info = next((u for u in users if u["email"] == user.email))
```
Comment (cscheng) [inline back/admin/integrations/sync_userinfo.py:31]: You could create a mapping dictionary outside the for loop to make it more efficient to lookup a user by email, e.g.: `users = {u.email: u for u in users}; users.get(email)`. There is also Django's `.in_bulk()` method: https://docs.djangoproject.com/en/4.2/ref/models/querysets/#in-bulk

**Review by cscheng — COMMENTED** [review]
This integration is very similar to the import users integration. In the manifest, `data_from` and `sync_data` accomplish almost the same thing. The main difference is that it's hard-coded to run daily. You could have it more generic by having a manifest where you can define:
* Whether to create new users or only update existing ones.
* Which fields to update
* The schedule of this integration, e.g. `once/daily/weekly`

The code for this integration could be contained in a single class. It depends on whether you want this flexibility or want the manifest a bit tidier. What are your thoughts?

File: back/admin/integrations/tasks.py [hunk 6]
```diff
@@ -1,9 +1,20 @@
...
+from admin.integrations.sync_userinfo import SyncUsers
 
 
 def retry_integration(new_hire_id, integration_id, params):
     integration = Integration.objects.get(id=integration_id)
     new_hire = get_user_model().objects.get(id=new_hire_id)
     integration.execute(new_hire, params)
+
+
+def sync_user_info(integration):
```
[the remark this answers is not in the document]
Comment (GDay) [reply back/admin/integrations/tasks.py]: Passing an object is fine for Django-Q2 (I maintain this library), but it looks like I have created a habit for this project to pass ids around, so I have changed it anyway!

**Comment by GDay** [comment]
I had considered merging the two, but since they serve very different purposes, I decided to split them.

However, having rewritten this pretty much from scratch, I do agree that this is likely the best approach. Thanks!

There is now a `schedule` option which needs a `cron` value. We likely don't need that much flexibility (hourly or daily would have been enough), but it's nice to have it anyway (since it's supported out of the box anyway and the validation is baked in).

**Comment by GDay** [comment]
By the way, I am super concerned about manifest validation/serialization at the moment. 
At some point I will make an interactive form where users can generate these manifest on the fly while also testing/debugging the requests.

File: back/admin/integrations/mixins.py [hunk 7]
```diff
@@ -121,22 +127,7 @@ def get_import_user_candidates(self, user):
             except KeyError:
                 break
 
-            users += self.extract_users_from_list_response(response)
+            users += self.extract_data_from_list_response(response)
```
Comment (cscheng) [inline back/admin/integrations/mixins.py:130]: The try-except above this line duplicates the initial behavior inside the `.extract_data_from_list_response()` method, could be simplified.

**Review by cscheng — APPROVED** [review]
This rewrite is very much an improvement over the previous two separate integrations. Looks and feels a lot cleaner!

While the new schedule option provides ultimate flexibility, keep in mind this could increase your application load unexpectedly if for example a user wants to update a lot of users every second. This might not necessarily be an issue if it's self-hosted or if the amount of users to be updated isn't huge, but it should be considered in the overall design of the application.
